### DISCUSSION TITLE
Install oh-my-posh for all Users

### DISCUSSION THREAD

**Discussion by nox309** [JanDeDobbeleer/oh-my-posh#3615] [Troubleshoot]
Hi, I may be blind, I am trying to install Oh-my-Posh for all users. I do the installation manually in a PowerShell script via PowerShell (manual installation). 
The installation works, but only for the admin user under which the script is executed. I have seen that in the Install.ps1 script, a switch is included to perform the installation for all users. But I can't find any info on how to do this.

**Comment by JanDeDobbeleer** [comment]
If you execute the script using `./install.ps1 -AllUsers` it will do that. Alternatively, you can use `winget` which has an `/ALLUSERS` switch.

**Reply by TiltonJH** [reply]
FYI: `winget install JanDeDobbeleer.OhMyPosh --scope machine`

`--scope machine` is the `/ALLUSERS` switch 😉

**Comment by nox309** [comment]
thank you, is it possible with the one-liner (Set-ExecutionPolicy Bypass -Scope Process -Force; Invoke-Expression ((New-Object System.Net.WebClient).DownloadString('https://ohmyposh.dev/install.ps1')) )? unfortunately Winget is blocked in our company network. And I would not like to download the install script and add it to the source of my project, which would mean that if the install script changes, I would not necessarily know about it.

**Reply by JanDeDobbeleer** [reply]
With the power of the interwebs:

```powershell
Invoke-Command -ScriptBlock ([scriptblock]::Create([System.Text.Encoding]::UTF8.GetString((New-Object Net.WebClient).DownloadData('https://ohmyposh.dev/install.ps1')))) -ArgumentList @('AllUsers')
```

**Reply by nox309** [reply]
Thanks, will test it

**Reply by biereluc** [reply]
@JanDeDobbeleer, is it possible that the syntax of your one-liner is wrong? You can't pass a switch parameter in a scriptblock. We need to pass the value of our argument, must **_$true_** to enable the `-AllUsers`. 

Like this :
```powershell
Invoke-Command -ScriptBlock ([scriptblock]::Create([System.Text.Encoding]::UTF8.GetString((New-Object Net.WebClient).DownloadData('https://ohmyposh.dev/install.ps1')))) -ArgumentList $true
```
More easy way:
```powershell
# With & call operator
& ([scriptblock]::Create([System.Text.Encoding]::UTF8.GetString((New-Object Net.WebClient).DownloadData('https://ohmyposh.dev/install.ps1')))) -AllUsers
# or with Invoke-WebRequest
& ([scriptblock]::Create((iwr 'https://ohmyposh.dev/install.ps1'))) -AllUsers
```

Anyway the following command throws an error in all the environments that I tried to run it.

```powershell
>Invoke-Command -ScriptBlock ([scriptblock]::Create([System.Text.Encoding]::UTF8.GetString((New-Object Net.WebClient).DownloadData('https://ohmyposh.dev/install.ps1')))) -ArgumentList @('AllUsers')
>Invoke-Command : Cannot convert the "AllUsers" value of type "System.String" to type
"System.Management.Automation.SwitchParameter".
At line:1 char:1
+ Invoke-Command -ScriptBlock ([scriptblock]::Create([System.Text.Encod ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Invoke-Command], PSInvalidCastException
    + FullyQualifiedErrorId : ConvertToFinalInvalidCastException,Microsoft.PowerShell.Commands.InvokeCommandCommand
```

**Reply by JanDeDobbeleer** [reply]
Could be, as I said, I just copy-pasted this.

**Comment by PennyCilin-gmx** [comment]
When I use the following script for installing ohmyposh, logged on as an admin starting Powershell 7.
`Invoke-Command -ScriptBlock ([scriptblock]::Create([System.Text.Encoding]::UTF8.GetString((New-Object Net.WebClient).DownloadData('https://ohmyposh.dev/install.ps1')))) -ArgumentList `$true``

It installs only the x86and not x64 for all users. I have PowerShell 7 (x64) installed.
`PS> (Get-Command oh-my-posh).Source
C:\Program Files (x86)\oh-my-posh\bin\oh-my-posh.exe`

What goes wrong?

And next problem is, update / upgrade with winget doesn't work.
How can I install OhMyPosh for all users and using winget for updates?

**Comment by JanDeDobbeleer** [comment]
@PennyCilin-gmx it's no longer possible to achieve this. I still need to refactor the install script to facilitate this however. We're moving to MSIX which doesn't offer system wide installation.

**Reply by PennyCilin-gmx** [reply]
@JanDeDobbeleer,

thank you very much for your reply. I logged on with an admin account. Downloaded the latest release of oh-my-posh (install-x64.msi) and installed it.
I check the Installation Path with `(Get-Command oh-my-posh).Source` and wondering:
the path is
 `C:\Program Files (x86)\oh-my-posh\bin\oh-my-posh.exe`
???
I remember the path for x64 executables where 
`C:\Program Files`
 
Okay, Oh-my-posh is available for all user accounts on my PC
I hope I can update this installation using winget in the future.

Greeting Penny.

**Reply by PennyCilin-gmx** [reply]
@Jan I respect your opinion. But I think for PCs which have some user accounts (two or more) it's easy to have a systemwide installation method, instead to do the install for every user.. But it can be also a security risk.

Greetings Penny.

**Reply by JanDeDobbeleer** [reply]
@PennyCilin-gmx it's not a personal decision, it's mostly due to MSIX not supporting this and MSI being a terrible experience.

**Reply by PennyCilin-gmx** [reply]
@JanDeDobbeleer 
Okay, I understand. But I'm a old school developer (long time ago), programming in Cobol, Fortran and Pascal. Scripting in REXX and learning more and more PowerShell.

Yes, I know programming in / for Windows is awful.
So now OhMyPosh works now on my Notebook with MS$ Windows 11 Pro.

Greetings Penny.

**Reply by PennyCilin-gmx** [reply]
Sorry that I ask again.
Today, winget shows that there is an update for oh my posh. I start PowerShell Core with an elevated prompt. perform

`winget update JanDeDobbeleer.OhMyPosh --scope machine --force`

and get the following message:

`Es wurde eine neuere Version gefunden, die Installationstechnologie unterscheidet sich jedoch von der aktuellen installierten Version. Deinstallieren Sie das Paket, und installieren Sie die neuere Version.`

So I cannot update Oh My Posh with winget for all users.

That's the problem, which is horrible for me. While I don't understand how to install Oh My Posh for all users on some PCs.

Is the best way, to use the following script, that I found in this thread?
`# With & call operator
& ([scriptblock]::Create([System.Text.Encoding]::UTF8.GetString((New-Object Net.WebClient).DownloadData('https://ohmyposh.dev/install.ps1')))) -AllUsers
# or with Invoke-WebRequest
& ([scriptblock]::Create((iwr 'https://ohmyposh.dev/install.ps1'))) -AllUsers`

Thank you very much for help and clarify this.

Greetings Penny.
